File: samlang-cli/src/main/kotlin/samlang/interpreter/AssemblyInterpreter.kt
package samlang.interpreter

import samlang.ast.asm.AssemblyArg
import samlang.ast.asm.AssemblyArgs
import samlang.ast.asm.AssemblyArgs.MEM
import samlang.ast.asm.AssemblyArgs.RAX
import samlang.ast.asm.AssemblyArgs.RBP
import samlang.ast.asm.AssemblyArgs.RDI
import samlang.ast.asm.AssemblyArgs.RDX
import samlang.ast.asm.AssemblyArgs.RSI
import samlang.ast.asm.AssemblyArgs.RSP
import samlang.ast.asm.AssemblyInstruction
import samlang.ast.asm.AssemblyInstruction.*
import samlang.ast.asm.AssemblyInstructionVisitor
import samlang.ast.asm.AssemblyProgram
import samlang.ast.asm.RegOrMem
import samlang.ast.common.IrNameEncoder

@ExperimentalStdlibApi
class AssemblyInterpreter(program: AssemblyProgram) {
    /** The list of all instructions. */
    private val instructions: List<AssemblyInstruction> = program.instructions
    /** The mapping between label and instruction number. Useful for jump. */
    private val labelInstructionNumberMapping: MutableMap<String, Int> = mutableMapOf()
    /** The mapping from names to actual memory address. */
    private val nameToMemoryAddress: MutableMap<String, Long> = mutableMapOf()
    /** Current register values. It will only be lazily provisioned. */
    private val registers: MutableMap<String, Long> = mutableMapOf()
    /** Current memory content. It will only be lazily provisioned. */
    private val memory: MutableMap<Long, Long> = mutableMapOf()
    /**
     * Current flags.
     * The flags are not exactly the same as the x86 ones.
     */
    private val flags: MutableMap<String, Boolean> = mutableMapOf()
    /** The current instruction pointer. */
    private var instructionPointer: Int
    /** The current heap end pointer. */
    private var currentHeapEndPointer: Int
    /** The place to collect all the stuff printed. */
    private val printCollector: StringBuilder
    /** The interpreter visitor to use.  */
    private val visitor: InterpreterVisitor

    init {
        var globalVarsTotalSize = 10000L
        for ((name, content) in program.globalVariables) {
            // Setup content variable size
            val contentStart = globalVarsTotalSize
            nameToMemoryAddress[name] = contentStart
            globalVarsTotalSize += content.length * 8 + 8
            // Setup content
            memory[contentStart] = content.length.toLong()
            val characterStart = contentStart + 8L
            content.toCharArray().forEachIndexed { index, character ->
                memory[characterStart + 8 * index] = character.toLong()
            }
        }
        // allocate space for global vars
        calloc(globalVarsTotalSize)
        val len = instructions.size
        for (i in 0 until len) {
            val instruction = instructions[i]
            if (instruction is Label) {
                labelInstructionNumberMapping[instruction.label] = i * 8
            }
        }
        registers[RSP.id] = 0x780000000L // set stack pointer
        instructionPointer = labelInstructionNumberMapping[IrNameEncoder.compiledProgramMain]!!
        currentHeapEndPointer = globalVarsTotalSize.toInt()
        printCollector = StringBuilder()
        visitor = InterpreterVisitor()
        stepUntilReturn()
    }

    /** @return the interpretation result string, which is the stdout result.  */
    val interpretationResult: String get() = printCollector.toString()

    private fun stepUntilReturn() {
        while (true) {
            if (instructionPointer % 8 != 0) {
                throw Error("BAD RIP: $instructionPointer")
            }
            val instruction = instructions[instructionPointer / 8]
            try {
                instruction.accept(visitor)
            } catch (e: ReturnException) {
                return
            }
            instructionPointer += 8
        }
    }

    /**
     * @param id the id of the register.
     * @return the value inside the given register, 0 if it's first used.
     */
    private fun getReg(id: String): Long {
        val value = registers[id]
        if (value != null) {
            return value
        }
        registers[id] = 0L
        return 0L
    }

    /**
     * @param id the id of the register.
     * @param value the value to store into the register.
     */
    private fun setReg(id: String, value: Long) {
        registers[id] = value
    }

    private fun checkMemLocation(location: Long) {
        if (location % 8 != 0L) {
            throw PanicException("Unaligned memory access: $location (word size=8)")
        }
        if (location < 0 || location > 0x780000000L) {
            throw PanicException("Segmentation fault: $location.")
        }
    }

    /**
     * @param location the memory location to access.
     * @return the value inside the memory location, 0 if first used.
     */
    private fun getMem(location: Long): Long {
        checkMemLocation(location = location)
        val value = memory[location]
        if (value != null) {
            return value
        }
        memory[location] = 0L
        return 0L
    }

    /**
     * @param location the memory location to set.
     * @param value the value to store into the memory location.
     */
    private fun setMem(location: Long, value: Long) {
        checkMemLocation(location = location)
        memory[location] = value
    }

    /**
     * @param mem the mem expression.
     * @return the value inside the memory location, 0 if first used.
     */
    private fun getMemLocation(mem: AssemblyArgs.Mem): Long {
        var memLoc: Long = 0
        val baseReg = mem.baseReg
        if (baseReg != null) {
            memLoc += getReg(baseReg.id)
        }
        val multipleOf = mem.multipleOf
        if (multipleOf != null) {
            val b = getReg(multipleOf.baseReg.id)
            val c = multipleOf.multipliedConstant.constant
            memLoc += b * c
        }
        val displacement = mem.displacement
        if (displacement != null) {
            memLoc += getConstValue(displacement)
        }
        return memLoc
    }

    private fun getConstValue(constant: AssemblyArgs.Const): Long {
        val value = constant.value
        val name = constant.name
        if (value == null && name == null || value != null && name != null) {
            throw Error("Corrupted: $constant")
        }
        if (value != null) {
            return value.toLong()
        }
        return nameToMemoryAddress[name] ?: labelInstructionNumberMapping[name]!!.toLong()
    }

    /**
     * @param arg the argument in an assembly instruction.
     * @return the stored value.
     */
    private fun getValue(arg: AssemblyArg): Long = arg.match(
        constF = { getConstValue(it) },
        regF = { (id) -> getReg(id) },
        memF = { mem -> getMem(getMemLocation(mem)) }
    )

    /**
     * @param regOrMem the argument in an assembly instruction.
     * @param value the value to set.
     */
    private fun setValue(regOrMem: RegOrMem, value: Long) {
        regOrMem.matchRegOrMem(
            regF = { (id) -> setReg(id = id, value = value) },
            memF = { mem -> setMem(location = getMemLocation(mem = mem), value = value) }
        )
    }

    /**
     * @param size size of the memory to allocate.
     * @return the allocated memory's starting pointer.
     */
    private fun calloc(size: Long): Long {
        if (size < 0) {
            throw PanicException("Invalid size")
        }
        if (size % 8 != 0L) {
            throw PanicException("Can only allocate in chunks of 8 bytes!: bad size: $size")
        }
        val pointerToBeReturned = currentHeapEndPointer.toLong()
        currentHeapEndPointer = (currentHeapEndPointer + size).toInt()
        if (currentHeapEndPointer > 1 shl 20) {
            throw PanicException("Out of heap!")
        }
        return pointerToBeReturned
    }

    /**
     * The hack to break the control flow and return from function call.
     */
    private class ReturnException : Error()

    private inner class InterpreterVisitor : AssemblyInstructionVisitor {
        override fun visit(node: MoveFromLong) {
            setValue(regOrMem = node.dest, value = node.value)
        }

        override fun visit(node: MoveToMem) {
            setValue(regOrMem = node.dest, value = getValue(arg = node.src))
        }

        override fun visit(node: MoveToReg) {
            setValue(regOrMem = node.dest, value = getValue(arg = node.src))
        }

        override fun visit(node: LoadEffectiveAddress) {
            setValue(regOrMem = node.dest, value = getMemLocation(mem = node.mem))
        }

        private fun visitCmp(minuend: AssemblyArg, subtrahend: AssemblyArg) {
            val m = getValue(arg = minuend)
            val s = getValue(arg = subtrahend)
            flags["eq"] = m == s
            flags["le"] = m <= s
            flags["lt"] = m < s
            flags["z"] = m - s == 0L
        }

        override fun visit(node: CmpConstOrReg) {
            visitCmp(minuend = node.minuend, subtrahend = node.subtrahend)
        }

        override fun visit(node: CmpMem) {
            visitCmp(minuend = node.minuend, subtrahend = node.subtrahend)
        }

        override fun visit(node: SetOnFlag) {
            val doesSetFlag = when (node.type) {
                JumpType.JMP -> throw Error()
                JumpType.JE -> flags["eq"]!!
                JumpType.JNE -> !flags["eq"]!!
                JumpType.JG -> !flags["le"]!!
                JumpType.JGE -> !flags["lt"]!!
                JumpType.JL -> flags["lt"]!!
                JumpType.JLE -> flags["le"]!!
                JumpType.JZ -> flags["z"]!!
                JumpType.JNZ -> !flags["z"]!!
            }
            setValue(node.reg, if (doesSetFlag) 1L else 0L)
        }

        override fun visit(node: JumpLabel) {
            val doesJump = when (node.type) {
                JumpType.JMP -> true
                JumpType.JE -> flags["eq"]!!
                JumpType.JNE -> !flags["eq"]!!
                JumpType.JG -> !flags["le"]!!
                JumpType.JGE -> !flags["lt"]!!
                JumpType.JL -> flags["lt"]!!
                JumpType.JLE -> flags["le"]!!
                JumpType.JZ -> flags["z"]!!
                JumpType.JNZ -> !flags["z"]!!
            }
            if (doesJump) {
                instructionPointer = labelInstructionNumberMapping[node.label]!!
            }
        }

        /**
         * @param arrayPointer the pointer to the array.
         * @return the string of the array.
         */
        private fun readArray(arrayPointer: Long): String {
            val len = getMem(location = arrayPointer - 8).toInt()
            val sb = StringBuilder()
            for (i in 0 until len) {
                sb.append(getMem(location = arrayPointer + i * 8).toInt().toChar())
            }
            return sb.toString()
        }

        override fun visit(node: CallAddress) {
            val savedInstructionPointer = instructionPointer
            val functionExpr = node.address
            if (functionExpr is AssemblyArgs.Const) {
                val (_, functionName) = functionExpr
                if (functionName != null) {
                    when (functionName) {
                        IrNameEncoder.nameOfPrintln -> {
                            val argument = getValue(arg = RDI)
                            printCollector.append(readArray(argument)).append('\n')
                            return
                        }
                        IrNameEncoder.nameOfIntToString -> {
                            val argument = getValue(arg = RDI)
                            val resultArray = argument.toString().toCharArray().map { it.toLong() }.toLongArray()
                            val memStartingPointer = calloc(resultArray.size * 8 + 8.toLong())
                            val unparsedStringStartingPointer = memStartingPointer + 8
                            setMem(memStartingPointer, resultArray.size.toLong())
                            var i = 0
                            while (i < resultArray.size) {
                                setMem(
                                    location = unparsedStringStartingPointer + i * 8,
                                    value = resultArray[i]
                                )
                                i++
                            }
                            setValue(RAX, unparsedStringStartingPointer)
                            return
                        }
                        IrNameEncoder.nameOfStringToInt -> {
                            val strToParse = readArray(getValue(RDI))
                            val value = strToParse.toLongOrNull() ?: throw PanicException("Bad string: $strToParse")
                            setValue(RAX, value)
                            return
                        }
                        IrNameEncoder.nameOfStringConcat -> {
                            val concatString = (readArray(getValue(RDI)) + readArray(getValue(RSI))).toCharArray()
                            val memStartingPointer = calloc(concatString.size * 8 + 8.toLong())
                            setMem(memStartingPointer, concatString.size.toLong())
                            val stringStartPointer = memStartingPointer + 8
                            for (i in concatString.indices) {
                                setMem(stringStartPointer + i * 8, concatString[i].toLong())
                            }
                            setValue(RAX, stringStartPointer)
                            return
                        }
                        IrNameEncoder.nameOfMalloc -> {
                            setValue(RAX, calloc(getValue(RDI)))
                            return
                        }
                        else -> {
                            instructionPointer = labelInstructionNumberMapping[functionName]!!
                            stepUntilReturn()
                            instructionPointer = savedInstructionPointer
                            return
                        }
                    }
                }
            }
            instructionPointer = getValue(functionExpr).toInt()
            stepUntilReturn()
            instructionPointer = savedInstructionPointer
        }

        override fun visit(node: Return) {
            throw ReturnException()
        }

        private fun interpretAlBinaryOpMemDest(
            dest: RegOrMem,
            src: AssemblyArg,
            type: AlBinaryOpType
        ) {
            val srcValue = getValue(arg = src)
            val destValue = getValue(arg = dest)
            val newValue = when (type) {
                AlBinaryOpType.ADD -> destValue + srcValue
                AlBinaryOpType.SUB -> destValue - srcValue
                AlBinaryOpType.XOR -> destValue xor srcValue
            }
            setValue(regOrMem = dest, value = newValue)
        }

        override fun visit(node: AlBinaryOpMemDest) {
            interpretAlBinaryOpMemDest(dest = node.dest, src = node.src, type = node.type)
        }

        override fun visit(node: AlBinaryOpRegDest) {
            interpretAlBinaryOpMemDest(dest = node.dest, src = node.src, type = node.type)
        }

        override fun visit(node: IMulTwoArgs) {
            setValue(
                regOrMem = node.dest,
                value = getValue(node.dest) * getValue(node.src)
            )
        }

        override fun visit(node: IMulThreeArgs) {
            setValue(
                regOrMem = node.dest,
                value = getValue(node.src) * getValue(node.immediate)
            )
        }

        override fun visit(node: Cqo) {
            if (getValue(RAX) >= 0) {
                setValue(regOrMem = RDX, value = 0)
            } else {
                setValue(regOrMem = RDX, value = -1)
            }
        }

        override fun visit(node: IDiv) {
            val raxValue = getValue(RAX)
            val argValue = getValue(node.divisor)
            if (argValue == 0L) {
                throw PanicException("Division by zero!")
            }
            val q = raxValue / argValue
            val r = raxValue % argValue
            setValue(RAX, q)
            setValue(RDX, r)
        }

        override fun visit(node: Neg) {
            setValue(node.dest, -getValue(node.dest))
        }

        override fun visit(node: ShiftLeft) {
            setValue(node.dest, getValue(node.dest) shl node.count)
        }

        override fun visit(node: Push) {
            val value = getValue(node.arg)
            setValue(regOrMem = RSP, value = getValue(RSP) - 8)
            setValue(MEM(reg = RSP), value)
        }

        override fun visit(node: PopRBP) {
            setValue(regOrMem = RBP, value = getValue(MEM(RSP)))
            setValue(regOrMem = RSP, value = getValue(RSP) + 8)
        }

        override fun visit(node: Label): Unit = Unit
        override fun visit(node: Comment): Unit = Unit
    }
}

File: samlang-core-interpreter/assembly-interpreter.ts
import PanicException from './panic-exception';

import {
  RAX,
  RDX,
  RSI,
  RDI,
  RSP,
  RBP,
  ASM_MEM_REG,
  AssemblyConst,
  AssemblyMemory,
  AssemblyRegisterOrMemory,
  AssemblyArgument,
} from 'samlang-core-ast/asm-arguments';
import type { AssemblyInstruction, AssemblyCall } from 'samlang-core-ast/asm-instructions';
import type { AssemblyProgram } from 'samlang-core-ast/asm-program';
import {
  ENCODED_COMPILED_PROGRAM_MAIN,
  ENCODED_FUNCTION_NAME_PRINTLN,
  ENCODED_FUNCTION_NAME_INT_TO_STRING,
  ENCODED_FUNCTION_NAME_STRING_TO_INT,
  ENCODED_FUNCTION_NAME_STRING_CONCAT,
  ENCODED_FUNCTION_NAME_MALLOC,
  ENCODED_FUNCTION_NAME_THROW,
} from 'samlang-core-ast/common-names';
import { Long, assertNotNull, checkNotNull } from 'samlang-core-utils';

// istanbul ignore next
const checkMemoryLocation = (location: Long): void => {
  if (location.mod(8).notEquals(Long.ZERO)) {
    throw new Error(`Unaligned memory access: ${location} (word size=8)`);
  }
  if (location.lessThan(0) || location.greaterThan(0x78000000)) {
    throw new Error(`Segmentation fault: ${location}.`);
  }
};

class ReturnException extends Error {}

class AssemblyInterpreter {
  /** The list of all instructions. */
  private readonly instructions: readonly AssemblyInstruction[];

  /** The mapping between label and instruction number. Useful for jump. */
  private readonly labelInstructionNumberMapping: Readonly<Record<string, number>>;

  /** The mapping from names to actual memory address. */
  private readonly nameToMemoryAddress: Readonly<Record<string, Long>>;

  /** Current register values. It will only be lazily provisioned. */
  private readonly registers: Record<string, Long> = {
    [RSP.id]: Long.fromInt(0x78000000),
  };

  /** Current memory content. It will only be lazily provisioned. */
  private readonly memory: Map<string, Long> = new Map();

  /**
   * Current flags.
   * The flags are not exactly the same as the x86 ones.
   */
  private readonly flags: Map<string, boolean> = new Map();

  /** The current instruction pointer. */
  private instructionPointer: number;

  /** The current heap end pointer. */
  private currentHeapEndPointer = 0;

  /** The place to collect all the stuff printed. */
  private printCollector = '';

  constructor(program: AssemblyProgram) {
    this.instructions = program.instructions;
    const labelInstructionNumberMapping: Record<string, number> = {};
    const nameToMemoryAddress: Record<string, Long> = {};
    let globalVarsTotalSize = Long.fromInt(10000);
    program.globalVariables.forEach(({ name, content }) => {
      // Setup content variable size
      const contentStart = globalVarsTotalSize;
      nameToMemoryAddress[name] = contentStart;
      globalVarsTotalSize = globalVarsTotalSize.add(content.length * 8 + 8);
      // Setup content
      this.memory.set(contentStart.toString(), Long.fromInt(content.length));
      const characterStart = contentStart.add(8);
      Array.from(content).forEach((characterString, index) => {
        this.memory.set(
          characterStart.add(8 * index).toString(),
          Long.fromInt(characterString.charCodeAt(0))
        );
      });
    });
    this.calloc(globalVarsTotalSize);
    program.instructions.forEach((instruction, index) => {
      if (instruction.__type__ === 'AssemblyLabel') {
        labelInstructionNumberMapping[instruction.label] = index * 8;
      }
    });
    this.labelInstructionNumberMapping = labelInstructionNumberMapping;
    this.nameToMemoryAddress = nameToMemoryAddress;
    const instructionPointer = labelInstructionNumberMapping[ENCODED_COMPILED_PROGRAM_MAIN];
    assertNotNull(instructionPointer);
    this.instructionPointer = instructionPointer;
    // istanbul ignore next
    if (this.currentHeapEndPointer !== Number(globalVarsTotalSize)) throw new Error();
    this.stepUntilReturn();
  }

  get interpretationResult(): string {
    return this.printCollector;
  }

  private stepUntilReturn(): void {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      // istanbul ignore next
      if (this.instructionPointer % 8 !== 0) throw new Error(`Bad RIP: ${this.instructionPointer}`);
      const index = this.instructionPointer / 8;
      const instruction = this.instructions[index];
      // istanbul ignore next
      if (instruction == null) throw new Error(`index=${index}, len=${this.instructions.length}`);
      try {
        this.interpret(instruction);
      } catch (e) {
        // istanbul ignore next
        if (e instanceof ReturnException) return;
        // istanbul ignore next
        throw e;
      }
      this.instructionPointer += 8;
    }
  }

  private getRegister = (id: string): Long => {
    const value = this.registers[id];
    // istanbul ignore next
    if (value != null) return value;
    // istanbul ignore next
    this.registers[id] = Long.ZERO;
    // istanbul ignore next
    return Long.ZERO;
  };

  private setRegister = (id: string, value: Long): void => {
    this.registers[id] = value;
  };

  private getMemory = (location: Long): Long => {
    checkMemoryLocation(location);
    const value = this.memory.get(location.toString());
    // istanbul ignore next
    if (value != null) return value;
    // istanbul ignore next
    this.memory.set(location.toString(), Long.ZERO);
    // istanbul ignore next
    return Long.ZERO;
  };

  private setMemory = (location: Long, value: Long): void => {
    checkMemoryLocation(location);
    this.memory.set(location.toString(), value);
  };

  private getConstValue = (constant: AssemblyConst): Long => {
    const valueOrName = constant.value;
    if (typeof valueOrName === 'number') return Long.fromInt(valueOrName);
    const nameValue = this.nameToMemoryAddress[valueOrName];
    if (nameValue != null) return nameValue;
    const instructionNumber = this.labelInstructionNumberMapping[valueOrName];
    assertNotNull(instructionNumber);
    return Long.fromInt(instructionNumber);
  };

  private getMemoryLocation = ({
    baseRegister,
    multipleOf,
    displacementConstant,
  }: AssemblyMemory): Long => {
    let memoryLocation = Long.ZERO;
    if (baseRegister != null) {
      memoryLocation = memoryLocation.add(this.getRegister(baseRegister.id));
    }
    if (multipleOf != null) {
      memoryLocation = memoryLocation.add(
        this.getRegister(multipleOf.baseRegister.id).multiply(multipleOf.multipliedConstant)
      );
    }
    if (displacementConstant != null) {
      memoryLocation = memoryLocation.add(this.getConstValue(displacementConstant));
    }
    return memoryLocation;
  };

  private getValue = (assemblyArgument: AssemblyArgument): Long => {
    switch (assemblyArgument.__type__) {
      case 'AssemblyConst':
        return this.getConstValue(assemblyArgument);
      case 'AssemblyRegister':
        return this.getRegister(assemblyArgument.id);
      case 'AssemblyMemory':
        return this.getMemory(this.getMemoryLocation(assemblyArgument));
    }
  };

  private setValue = (target: AssemblyRegisterOrMemory, value: Long): void => {
    switch (target.__type__) {
      case 'AssemblyRegister':
        this.setRegister(target.id, value);
        break;
      case 'AssemblyMemory':
        this.setMemory(this.getMemoryLocation(target), value);
        break;
    }
  };

  private calloc = (size: Long): Long => {
    // istanbul ignore next
    if (size.lessThan(Long.ZERO)) throw new Error('Invalid size');
    // istanbul ignore next
    if (size.mod(8).notEquals(Long.ZERO)) {
      // istanbul ignore next
      throw new Error(`Can only allocate in chunks of 8 bytes!: bad size: ${size}`);
    }
    const pointerToBeReturned = Long.fromInt(this.currentHeapEndPointer);
    this.currentHeapEndPointer += Number(size);
    return pointerToBeReturned;
  };

  private interpret = (node: AssemblyInstruction): void => {
    switch (node.__type__) {
      // istanbul ignore next
      case 'AssemblyMoveFromLong':
        // istanbul ignore next
        this.setValue(node.destination, node.value);
        return;
      case 'AssemblyMoveToMemory':
      case 'AssemblyMoveToRegister':
        this.setValue(node.destination, this.getValue(node.source));
        return;
      case 'AssemblyLoadEffectiveAddress':
        this.setValue(node.destination, this.getMemoryLocation(node.source));
        return;
      // istanbul ignore next
      case 'AssemblyCompareMemory':
      case 'AssemblyCompareConstOrRegister': {
        const m = this.getValue(node.minuend);
        const s = this.getValue(node.subtrahend);
        this.flags.set('eq', m.equals(s));
        this.flags.set('le', m.lessThanOrEqual(s));
        this.flags.set('lt', m.lessThan(s));
        this.flags.set('z', m.subtract(s).equals(Long.ZERO));
        return;
      }
      case 'AssemblySetOnFlag': {
        let doesSetFlag: boolean | undefined;
        // istanbul ignore next
        switch (node.type) {
          case 'je':
            doesSetFlag = this.flags.get('eq');
            break;
          case 'jne':
            doesSetFlag = !this.flags.get('eq');
            break;
          case 'jl':
            doesSetFlag = this.flags.get('lt');
            break;
          case 'jle':
            doesSetFlag = this.flags.get('le');
            break;
          case 'jg':
            doesSetFlag = !this.flags.get('le');
            break;
          case 'jge':
            doesSetFlag = !this.flags.get('lt');
            break;
          case 'jz':
            doesSetFlag = this.flags.get('z');
            break;
          case 'jnz':
            doesSetFlag = !this.flags.get('z');
            break;
        }
        assertNotNull(doesSetFlag);
        // istanbul ignore next
        this.setValue(node.register, doesSetFlag ? Long.ONE : Long.ZERO);
        return;
      }
      case 'AssemblyJump': {
        let doesJump: boolean | undefined;
        // istanbul ignore next
        switch (node.type) {
          case 'jmp':
            doesJump = true;
            break;
          case 'je':
            doesJump = this.flags.get('eq');
            break;
          case 'jne':
            doesJump = !this.flags.get('eq');
            break;
          case 'jl':
            doesJump = this.flags.get('lt');
            break;
          case 'jle':
            doesJump = this.flags.get('le');
            break;
          case 'jg':
            doesJump = !this.flags.get('le');
            break;
          case 'jge':
            doesJump = !this.flags.get('lt');
            break;
          case 'jz':
            doesJump = this.flags.get('z');
            break;
          case 'jnz':
            doesJump = !this.flags.get('z');
            break;
        }
        assertNotNull(doesJump);
        if (doesJump) {
          const newPointer = this.labelInstructionNumberMapping[node.label];
          assertNotNull(newPointer);
          this.instructionPointer = newPointer;
        }
        return;
      }
      case 'AssemblyCall':
        this.interpretCall(node);
        return;
      case 'AssemblyReturn':
        throw new ReturnException();
      // istanbul ignore next
      case 'AssemblyArithmeticBinaryMemoryDestination':
      case 'AssemblyArithmeticBinaryRegisterDestination': {
        const sourceValue = this.getValue(node.source);
        const destinationValue = this.getValue(node.destination);
        let newValue: Long;
        switch (node.type) {
          case 'add':
            newValue = destinationValue.add(sourceValue);
            break;
          case 'sub':
            newValue = destinationValue.subtract(sourceValue);
            break;
          case 'xor':
            newValue = destinationValue.xor(sourceValue);
            break;
        }
        this.setValue(node.destination, newValue);
        return;
      }
      case 'AssemblyIMulTwoArgs':
        this.setValue(
          node.destination,
          this.getValue(node.destination).multiply(this.getValue(node.source))
        );
        return;
      // istanbul ignore next
      case 'AssemblyIMulThreeArgs':
        this.setValue(
          node.destination,
          this.getValue(node.source).multiply(this.getValue(node.immediate))
        );
        return;
      // istanbul ignore next
      case 'AssemblyCqo':
        if (this.getValue(RAX).greaterThanOrEqual(Long.ZERO)) {
          this.setValue(RDX, Long.ZERO);
        } else {
          this.setValue(RDX, Long.NEG_ONE);
        }
        return;
      case 'AssemblyIDiv': {
        const raxValue = this.getValue(RAX);
        const argumentValue = this.getValue(node.divisor);
        // istanbul ignore next
        if (argumentValue.equals(Long.ZERO)) throw new PanicException('Division by zero!');
        this.setValue(RAX, raxValue.divide(argumentValue));
        this.setValue(RDX, raxValue.mod(argumentValue));
        return;
      }
      // istanbul ignore next
      case 'AssemblyNeg':
        this.setValue(node.destination, this.getValue(node.destination).negate());
        return;
      // istanbul ignore next
      case 'AssemblyShiftLeft':
        // eslint-disable-next-line no-bitwise
        this.setValue(node.destination, this.getValue(node.destination).shiftLeft(node.count));
        return;
      case 'AssemblyPush': {
        const value = this.getValue(node.pushArgument);
        this.setValue(RSP, this.getValue(RSP).subtract(8));
        this.setValue(ASM_MEM_REG(RSP), value);
        return;
      }
      case 'AssemblyPopRBP':
        this.setValue(RBP, this.getValue(ASM_MEM_REG(RSP)));
        this.setValue(RSP, this.getValue(RSP).add(8));
        break;
      case 'AssemblyLabel':
      case 'AssemblyComment':
    }
  };

  private readArray = (arrayPointer: Long): string => {
    const len = Number(this.getMemory(arrayPointer));
    const array: number[] = [];
    for (let i = 1; i <= len; i += 1) {
      array.push(Number(this.getMemory(arrayPointer.add(i * 8))));
    }
    return String.fromCharCode(...array);
  };

  private interpretCall = (node: AssemblyCall): void => {
    const savedInstructionPointer = this.instructionPointer;
    const functionExpression = node.address;
    if (
      functionExpression.__type__ === 'AssemblyConst' &&
      typeof functionExpression.value === 'string'
    ) {
      const functionName = functionExpression.value;
      switch (functionName) {
        case ENCODED_FUNCTION_NAME_PRINTLN: {
          const argument = this.getValue(RDI);
          this.printCollector += `${this.readArray(argument)}\n`;
          return;
        }
        case ENCODED_FUNCTION_NAME_INT_TO_STRING: {
          const argument = this.getValue(RDI);
          const resultArray = Array.from(String(argument)).map((it) =>
            Long.fromInt(it.charCodeAt(0))
          );
          const memoryStartPointer = this.calloc(Long.fromInt(resultArray.length * 8 + 8));
          const unparsedStringStartingPointer = memoryStartPointer.add(8);
          this.setMemory(memoryStartPointer, Long.fromInt(resultArray.length));
          for (let i = 0; i < resultArray.length; i += 1) {
            this.setMemory(unparsedStringStartingPointer.add(i * 8), checkNotNull(resultArray[i]));
          }
          this.setValue(RAX, memoryStartPointer);
          return;
        }
        case ENCODED_FUNCTION_NAME_STRING_TO_INT: {
          const stringToParse = this.readArray(this.getValue(RDI));
          try {
            BigInt(stringToParse);
            this.setValue(RAX, Long.fromString(stringToParse));
            return;
            // istanbul ignore next
          } catch {
            // istanbul ignore next
            throw new PanicException(`Bad string: ${stringToParse}`);
          }
        }
        case ENCODED_FUNCTION_NAME_STRING_CONCAT: {
          const concatString = Array.from(
            this.readArray(this.getValue(RDI)) + this.readArray(this.getValue(RSI))
          ).map((it) => Long.fromInt(it.charCodeAt(0)));
          const memoryStartPointer = this.calloc(Long.fromInt(concatString.length * 8 + 8));
          this.setMemory(memoryStartPointer, Long.fromInt(concatString.length));
          const stringStartPointer = memoryStartPointer.add(8);
          for (let i = 0; i < concatString.length; i += 1) {
            this.setMemory(stringStartPointer.add(i * 8), checkNotNull(concatString[i]));
          }
          this.setValue(RAX, memoryStartPointer);
          return;
        }
        case ENCODED_FUNCTION_NAME_MALLOC:
          this.setValue(RAX, this.calloc(this.getValue(RDI)));
          return;
        // istanbul ignore next
        case ENCODED_FUNCTION_NAME_THROW:
          throw new PanicException(this.readArray(this.getValue(RDI)));
        default: {
          const newInstructionPointer = this.labelInstructionNumberMapping[functionName];
          assertNotNull(newInstructionPointer);
          this.instructionPointer = newInstructionPointer;
          this.stepUntilReturn();
          this.instructionPointer = savedInstructionPointer;
          return;
        }
      }
    }
    this.instructionPointer = Number(this.getValue(functionExpression));
    this.stepUntilReturn();
    this.instructionPointer = savedInstructionPointer;
  };
}

const interpretAssemblyProgram = (program: AssemblyProgram): string =>
  new AssemblyInterpreter(program).interpretationResult;

export default interpretAssemblyProgram;
